packages/block-editor/src/components/global-styles/use-global-styles-output.js

```diff
@@ -830,39 +836,34 @@ export const toStyles = (
 							// Otherwise add regular selectors.
 							const styleVariationDeclarations =
 								getStylesDeclarations(
-									styles?.variations?.[ styleVariationName ],
+									styleVariations,
 									styleVariationSelector,
 									useRootPaddingAlign,
 									tree
 								);
-							if ( !! styleVariationDeclarations.length ) {
-								ruleset =
-									ruleset +
-									`${ styleVariationSelector }{${ styleVariationDeclarations.join(
-										';'
-									) }}`;
+							if ( styleVariationDeclarations.length ) {
+								ruleset += `${ styleVariationSelector }{${ styleVariationDeclarations.join(
+									';'
+								) };}`;
 							}
 						}
 					}
 				);
 			}
 
-			const duotoneStyles = {};
-			if ( styles?.filter ) {
-				duotoneStyles.filter = styles.filter;
-				delete styles.filter;
-			}
-
 			// Process duotone styles.
 			if ( duotoneSelector ) {
+				const duotoneStyles = {};
+				if ( styles?.filter ) {
+					duotoneStyles.filter = styles.filter;
+					delete styles.filter;
```

Comment:
Isn't there another fix where we avoid "delete" (avoid mutations at all)? It feels to me like if we keep it around, it can regress again at anytime with some small unintended changes.

Reply:
That would be very desirable and I spent some time thinking about how to do it. It wouldn't be a "fix" but rather a major rewrite of the `toStyles` function.

Currently it processes feature selectors and the root selector in separate loops. Pseudocode:
```
for ( featSelector of featureSelectors ) {
  declarations = [];
  for ( style of styles ) {
    if ( styleUsesSelector( style, featSelector ) ) {
      declarations.push( style );
      delete styles[ style ];
    }
  }
  write( `${ featSelector } { ${ declarations.join() } }`;
}

declarations = [];
for ( style of styles ) {
  declarations.push( style );
}
write( `${ rootSelector } { ${ declarations.join() } }`;
```
The second loop processes only styles that haven't been deleted by the first loop.

Rewritten version would have one loop, assigning each style to the right selector:
```
declarations = new Map<string, string[]>();
for ( style of styles ) {
  selector = findFeatureSelector( style ) || rootSelector;
  declarations.get( selector ).push( style );
}
for ( [ selector, decls ] of declarations ) {
  write( `${ selector } { ${ decls.join() } }` );
}
```
No `delete` needed.

Consider this PR a first step towards that goal. It's adding many new unit tests, which will be needed to perform the rewrite safely.